container/memory_store.go

```diff
@@ -1,67 +1,78 @@
 package container
 
-import "sync"
+import (
+	"sync"
+
+	"github.com/docker/docker/pkg/locker"
+)
 
 // memoryStore implements a Store in memory.
 type memoryStore struct {
-	s map[string]*Container
-	sync.Mutex
+	s          map[string]*Container
+	globalLock sync.Mutex
```

Comment:
I'm not sure where else to put it, but maybe leave a comment about `acquire the globallock before the namedlock`.